### PR TITLE
Add arduino-cli Formula

### PR DESCRIPTION
**Pull request by SConaway**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

I'm confused on how to build something with `go` so am looking for a step in the correct direction.

### CODE REVIEW

**Comment by SMillerDev** [comment]
There's a lot of formulae that build go. I think the easiest thing is just searching the repository for `depends_on "go"`

**Comment by SConaway** [comment]
So would my current `def install` work?

**Comment by SMillerDev** [comment]
Jenkins says it doesn't. You can try by running `brew install --build-from-source <formula>`

**Comment by SConaway** [comment]
That seems to work on my mac.

**Comment by SConaway** [comment]
Can someone help me?

File: Formula/arduino-cli.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    (buildpath/"src/github.com/arduino/arduino-cli").install buildpath.children
+
+    cd "src/github.com/arduino/arduino-cli" do
+      system "go", "build", "-o", bin/"arduino-cli"
+      prefix.install_metafiles
+    end
+  end
+
+  test do
+    system bin/"arduino-cli", "version"
```
Comment (igas) [inline Formula/arduino-cli.rb]: Can we test more functionality of the app? From the [docs](https://docs.brew.sh/Formula-Cookbook#add-a-test-to-the-formula):

> We want tests that don’t require any user input and test the basic functionality of the application. For example `foo build-foo input.foo` is a good test and (despite their widespread use) `foo --version` and `foo --help` are bad tests. However, a bad test is better than no test at all.

**Comment by igas** [comment]
Apart from test part this LGTM

File: Formula/arduino-cli.rb (hunk 1 above)
Comment (SConaway) [reply Formula/arduino-cli.rb]: Yes, but they will require installation of more components (Board packages). I can do that, but I'm not sure that's what you want.

Comment (SConaway) [reply Formula/arduino-cli.rb]: @igas Hello?

**Comment by SConaway** [comment]
Sorry about all the commits

File: Formula/arduino-cli.rb (hunk 1 above)
Comment (igas) [reply Formula/arduino-cli.rb]: Sorry for the delay. I can see in readme:
```
$ arduino-cli sketch new MyFirstSketch
Sketch created in: /home/luca/Arduino/MyFirstSketch

$ cat /home/luca/Arduino/MyFirstSketch/MyFirstSketch.ino
void setup() {
}

void loop() {
}
```
We can do the same, assert that `sketch new` have output `Sketch created` and that `MyFirstSketch/MyFirstSketch.ino` equal this content.

**Comment by SConaway** [comment]
@igorbolic This doesn't quite work. I can't figure out how to make the cli use the current dir.

````
$ brew test ./arduino-cli.rb
Testing arduino-cli
==> /usr/local/Cellar/arduino-cli/0.3.3-alpha/bin/arduino-cli sketch new TestSke
Last 15 lines from /Users/steven/Library/Logs/Homebrew/arduino-cli/test.01.arduino-cli:
2018-12-20 16:41:39 -0800

/usr/local/Cellar/arduino-cli/0.3.3-alpha/bin/arduino-cli
sketch
new
TestSketch

Error: mkdir /Users/steven/Documents/Arduino/TestSketch: operation not permitted
Could not create sketch directory.
Error: arduino-cli: failed
Failed executing:
/usr/local/Homebrew/Library/Homebrew/formula.rb:1921:in `block in system'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1858:in `open'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1858:in `system'
/Users/steven/homebrew-core/Formula/arduino-cli.rb:21:in `block in <class:ArduinoCli>'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1732:in `block (3 levels) in run_test'
/usr/local/Homebrew/Library/Homebrew/utils.rb:523:in `with_env'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1731:in `block (2 levels) in run_test'
/usr/local/Homebrew/Library/Homebrew/formula.rb:870:in `with_logging'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1730:in `block in run_test'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1971:in `block in mktemp'
/usr/local/Homebrew/Library/Homebrew/mktemp.rb:55:in `block in run'
/usr/local/Homebrew/Library/Homebrew/mktemp.rb:55:in `chdir'
/usr/local/Homebrew/Library/Homebrew/mktemp.rb:55:in `run'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1970:in `mktemp'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1724:in `run_test'
/usr/local/Homebrew/Library/Homebrew/test.rb:28:in `block in <main>'
/System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/timeout.rb:91:in `block in timeout'
/System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/timeout.rb:33:in `block in catch'
/System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/timeout.rb:33:in `catch'
/System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/timeout.rb:33:in `catch'
/System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/timeout.rb:106:in `timeout'
/usr/local/Homebrew/Library/Homebrew/test.rb:27:in `<main>'
````

**Comment by igas** [comment]
I think you need to create `.cli-config.yml`:

```
- paths:
  - sketchbook_path: #{buildpath}
```

**Comment by SConaway** [comment]
@igas I'm 100% stuck. Here's what I have:

````
test do
    config = File.open(".cli-config.yml", "w")
    config.write <<~EOS
    - paths:
    - sketchbook_path: #{buildpath}
    EOS
    system "#{bin}/arduino-cli", "sketch", "new", "TestSketch"
    code = File.open("TestSketch/TestSketch.ino", "r")
    assert_equal "void setup() {
}

void loop() {
}
", code
  end
````

**Comment by SConaway** [comment]
@igas can I get some help with the test?

**Comment by claui** [comment]
@SConaway Looks like [an upstream issue](https://github.com/arduino/arduino-cli/issues/30) to me. Expecting a user-level configuration file to be in `bin` is problematic: doing so not only violates the [FHS](https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard) but also clashes with how Homebrew handles kegs when upgrading.

Looking through the upstream issue, it seems to me that no one has decided yet as to which part of `.cli-config.yml` is considered user-level configuration after all, and in what directories that part of the file is supposed to reside.

With all that in mind, I feel that maybe it’s best to defer accepting this formula into homebrew-core until upstream has sorted out the issue.

**Comment by SConaway** [comment]
@claui should the test just be `arduino-cli --version` for now? I'll open a PR on upstream.

**Comment by claui** [comment]
@SConaway I’m leaning towards leaving this PR open and wait for the upstream fix before we merge.

I feel having a formula whose only config file gets clobbered by Homebrew on every update would be poor UX.

Pinging @Homebrew/maintainers for a second opinion.

**Comment by SMillerDev** [comment]
Sounds like a good option @claui

**Comment by fxcoudert** [comment]
No sign of activity, closing for now.

**Pull request closed without merging**